File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/component/IMU.java
package org.firstinspires.ftc.teamcode.component;


import com.qualcomm.hardware.rev.RevHubOrientationOnRobot.LogoFacingDirection;
import com.qualcomm.hardware.rev.RevHubOrientationOnRobot.UsbFacingDirection;
import com.qualcomm.hardware.rev.RevHubOrientationOnRobot;
import com.qualcomm.robotcore.hardware.HardwareMap;

import org.firstinspires.ftc.robotcore.external.navigation.AngleUnit;
import org.firstinspires.ftc.robotcore.external.navigation.Orientation;

public class IMU {
    public static AngleUnit DEGREES = AngleUnit.DEGREES;
    public static AngleUnit RADIANS = AngleUnit.RADIANS;

    public static LogoFacingDirection[] logoDirections = {
            LogoFacingDirection.UP,
            LogoFacingDirection.DOWN,
            LogoFacingDirection.LEFT,
            LogoFacingDirection.RIGHT,
            LogoFacingDirection.FORWARD,
            LogoFacingDirection.BACKWARD};
    public static UsbFacingDirection[] USBDirections = {
            UsbFacingDirection.UP,
            UsbFacingDirection.DOWN,
            UsbFacingDirection.LEFT,
            UsbFacingDirection.RIGHT,
            UsbFacingDirection.FORWARD,
            UsbFacingDirection.BACKWARD};
    public static int UP      = 0;
    public static int DOWN    = 1;
    public static int LEFT    = 2;
    public static int RIGHT   = 3;
    public static int FORWARD = 4;
    public static int BACKWARD = 5;
    private com.qualcomm.robotcore.hardware.IMU imu;
    private AngleUnit unit = RADIANS;
    private double offset = 0;

    public IMU(String name, HardwareMap hardwareMap){
        this.imu = hardwareMap.get(com.qualcomm.robotcore.hardware.IMU.class, name);
    }

    public void setUnit(AngleUnit unit){
        if(unit == this.unit){
            return;
        }
        this.unit = unit;
        if(unit == DEGREES){
            offset = offset * 180 / Math.PI;
        }
        else{
            offset = offset / 180 * Math.PI;
        }
    }
    public void configureOrientation(int logoDirection, int USBDirection){
        imu.initialize(
                new com.qualcomm.robotcore.hardware.IMU.Parameters(new RevHubOrientationOnRobot(
                        logoDirections[logoDirection],
                        USBDirections[USBDirection]
                ))
        );
    }

    public double getPich(){
        return imu.getRobotYawPitchRollAngles().getPitch(unit);
    }
    public double getRoll(){
        return imu.getRobotYawPitchRollAngles().getRoll(unit);
    }
    public double getYaw(){
        return imu.getRobotYawPitchRollAngles().getYaw(unit) + offset;
    }
    public void resetYaw(double angle){
        imu.resetYaw();
        offset += angle;
    }
    public void resetYaw(){
        imu.resetYaw();
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/component/IMU.kt
package org.firstinspires.ftc.teamcode.component

import com.qualcomm.hardware.rev.RevHubOrientationOnRobot
import com.qualcomm.hardware.rev.RevHubOrientationOnRobot.LogoFacingDirection
import com.qualcomm.hardware.rev.RevHubOrientationOnRobot.UsbFacingDirection
import com.qualcomm.robotcore.hardware.IMU
import org.firstinspires.ftc.robotcore.external.navigation.AngleUnit
import org.firstinspires.ftc.teamcode.hardware.HardwareMap
import org.firstinspires.ftc.teamcode.util.geometry.Rotation2D

class IMU(name: String) {
    private val imu = HardwareMap.get(IMU::class.java, name)
    private val unit = RADIANS

    private var offset = Rotation2D()
    var yaw = Rotation2D()
    var lastYaw = Rotation2D()

    fun configureOrientation(logo: Direction, usb: Direction) {
        imu.initialize(
            IMU.Parameters(RevHubOrientationOnRobot(
                logoDirections[logo],
                USBDirections[usb]
            ))
        )
    }

    val pitch: Rotation2D
        get() = Rotation2D(imu.robotYawPitchRollAngles.getPitch(unit))
    val roll: Rotation2D
        get() = Rotation2D(imu.robotYawPitchRollAngles.getRoll(unit))
    val delta: Rotation2D
        get() = (yaw - lastYaw).wrap()

    fun update(){
        lastYaw = yaw
        yaw = Rotation2D(imu.robotYawPitchRollAngles.getYaw(unit)) + offset
    }
    fun resetYaw(){
        offset = -Rotation2D(imu.robotYawPitchRollAngles.getYaw(unit))
    }

    enum class Direction {
        UP, DOWN, LEFT, RIGHT, FORWARD, BACKWARD
    }

    companion object {
        var RADIANS = AngleUnit.RADIANS
        var logoDirections = mapOf(
            Direction.UP       to LogoFacingDirection.UP,
            Direction.DOWN     to LogoFacingDirection.DOWN,
            Direction.LEFT     to LogoFacingDirection.LEFT,
            Direction.RIGHT    to LogoFacingDirection.RIGHT,
            Direction.FORWARD  to LogoFacingDirection.FORWARD,
            Direction.BACKWARD to LogoFacingDirection.BACKWARD)
        var USBDirections = mapOf(
            Direction.UP       to UsbFacingDirection.UP,
            Direction.DOWN     to UsbFacingDirection.DOWN,
            Direction.LEFT     to UsbFacingDirection.LEFT,
            Direction.RIGHT    to UsbFacingDirection.RIGHT,
            Direction.FORWARD  to UsbFacingDirection.FORWARD,
            Direction.BACKWARD to UsbFacingDirection.BACKWARD)
    }
}
